common/network-types/src/frame.rs

```diff
@@ -320,15 +365,23 @@ impl<'a> FrameReassembler<'a> {
     }
 
     /// Pushes a new [Segment] for reassembly.
-    pub fn push_segment(&self, segment: Segment<'a>) -> crate::errors::Result<()> {
+    /// This function also pushes out the reassembled frame if this segment completed it.
+    /// Pushing a segment belonging to a frame ID that has been already
+    /// previously completed or [evicted](FrameReassembler::evict) will fail.
+    pub fn push_segment(&self, segment: Segment) -> crate::errors::Result<()> {
         if self.reassembled.is_closed() {
-            return Err(ReassemblerClosed);
+            return Err(NetworkTypeError::ReassemblerClosed);
+        }
+
+        // Check if this frame has not been emitted yet.
+        let frame_id = segment.frame_id;
+        if frame_id <= self.last_emitted_frame.load(Ordering::SeqCst) {
+            return Err(NetworkTypeError::OldSegment(frame_id));
         }
```

Comment:
Why couldn't the frames be out of order though? A generic reassembler would not correlate individual frames, would it?

Reply:
As in the docs: The reassembler can maintain the order as long as all the segments of frame N+1 don't arrive earlier than all segments of frame N. In this case in hard drops the frame, to avoid out of sequence frame being emitted. 
This needs to be avoided by choosing a suitable frame size.

Reply:
It would be probably better named Frame Reassembler & Sequencer

Reply:
Alternatively, we can move the sequencing at an upper layer in SessionState. We'll see whats more appropriate in the wild.